### DISCUSSION TITLE
Python Docs Request

### DISCUSSION THREAD

**Discussion by photkey** [Flow-Launcher/Flow.Launcher#832] [General]
自学Python的新手，只看官方提供的例子，实在是看不太懂，现在急需开发文档。

用Python实现的一个简单功能，自认为非常的好用，比官方的网页搜索插件更好用，在官方现有功能的基础上，还可以:
批量搜索，比方同时 从百度和谷歌搜索；
输入中文自动翻译成英文， 然后打开搜索；
输入中文 从中文网站搜索中文，从英文网站搜索翻译后的英文，同时批量打开搜索页面 等等

但是现在不是以插件的形式运行，而是通过命令行插件运行，使用体验大打折扣，比方输入的内容，如果含有空格，必须还要加上引号。

所以现在非常希望能有一份详细的python插件开发文档
。

**Comment by taooceros** [comment]
实现插件的思路很简单，从命令行参数中读取query的信息，然后处理完之后以json的形式print到输出流。
格式参考https://github.com/Flow-Launcher/Flow.Launcher.JsonRPC.Python/blob/master/flowlauncher/FlowLauncher.py

或者你也可以直接下载FlowLauncher包，然后参考https://github.com/Flow-Launcher/Flow.Launcher.JsonRPC.Python的教程搭建。

插件的文档我们也希望能尽快实现一份完整的，不过文档耗时耗力，如果你愿意的话可以在弄明白之后帮忙改

**Reply by taooceros** [reply]
参考这个也行，最简单的实现，但几乎没有处理特殊情况和回调
```python
# -*- coding: utf-8 -*-

import json
import time
import random
import sys

time.sleep(1)

args = json.loads(sys.argv[1])

setting = args['settings']

results = \
    {
        "result": [
            {
                "Title": f"test",
                "SubTitle": "This is my hello world",
                "JsonRPCAction": {
                    'method': 'Flow.Launcher.ShowMsgError',
                    'parameters': ["test", 'test'],
                },

            },
            {
                "Title": f"{setting}",
                "SubTitle": "This is my hello world",
                "JsonRPCAction": {
                    'method': 'Flow.Launcher.ShowMsgError',
                    'parameters': ["test", 'test'],
                }
            }
        ],
        "SettingsChange": {
            "TestField": random.randint(1, 100),
            "UserName": str(random.randint(1, 100)),
            "Reason": str(random.randint(1, 100)),
            "Code Conduct": False,
            "dropdown": "test3"
        }
    }

print(json.dumps(results))
```

**Reply by photkey** [reply]
我非常想给这个项目做一点贡献，但是以我的能力写文档的话，好像是有点费力，二把刀的新手。
先看一下你刚发过来的链接，希望能看明白。其实我大概可以明白他的工作流程是怎么样的，就是具体应该怎么写，没搞明白。

我现在希望是能先把这个插件搞定，对于像我一样英文很差，又正在自学编程的人，应该会比较有用。

**Reply by photkey** [reply]
我就需要一个这样简单的，因为不需要回调。

**Reply by taooceros** [reply]
> 我就需要一个这样简单的，因为不需要回调。

你需要回调的，但可以以比较简单的方式实现。不然没办法在输入完搜索内容之后再进行搜索。这一段仅仅是让flow可以展示一部分的结果。`  'method': 'Flow.Launcher.ShowMsgError',` 这里是访问flow的api，具体选项参考https://flow-launcher.github.io/docs/#/json-rpc

https://github.com/Flow-Launcher/Flow.Launcher.Plugin.PythonTemplate 话说这个模板你试过了么

**Reply by photkey** [reply]
我以为回调指的是  Python脚本运行的结果传给 flow 叫回调……

看过，看过之后 我甚至不知道应该把我Python的代码写在哪个文件里面，纯纯的小白，让你见笑了。

**Reply by taooceros** [reply]
> 我以为回调指的是 Python脚本运行的结果传给 flow 叫回调……
> 
> 看过，看过之后 我甚至不知道应该把我Python的代码写在哪个文件里面，纯纯的小白，让你见笑了。

没事大家都是这样过来的，你试一下那个template呗，虽然貌似他可能不兼容最新flow会装的那个portable的python，但我相信你电脑里有python

**Reply by taooceros** [reply]
虽然但是，如果可以的话我们鼓励你用英文提问用英文回答。

**Reply by photkey** [reply]
好，谢谢你的帮助和建议！下次一定 😀
